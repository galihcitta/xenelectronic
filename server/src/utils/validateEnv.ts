import { port, cleanEnv } from 'envalid';

export const validateEnv = () => {
  cleanEnv(process.env, {
    PORT: port()
  });
};

// https://www.npmjs.com/package/envalid
