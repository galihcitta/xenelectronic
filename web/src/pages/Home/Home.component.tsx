import React from 'react';
import { connect } from 'react-redux';

import {
  Wrapper,
  NavLink,
  NavTitle,
  NavImage,
  ImageWrapper
} from './Home.styles';

import { fetchProductsByType } from '../../modules/shop/shop.actions';
import { Type } from '../../modules/shop/shop.interfaces';

import laptopImg from '../../assets/laptop.jpg';
import phoneImg from '../../assets/hp.jpg';
import acImg from '../../assets/ac.jpg';
import tvImg from '../../assets/tv.jpg';
import vcImg from '../../assets/vc.jpg';
import wmImg from '../../assets/wm.jpg';

const categories = [
  {
    title: 'SmartPhone',
    src: phoneImg
  },
  {
    title: 'TV',
    src: tvImg
  },
  {
    title: 'Laptop',
    src: laptopImg
  },
  {
    title: 'AC',
    src: acImg
  },
  {
    title: 'VacuumCleaner',
    src: vcImg
  },
  {
    title: 'WashMachine',
    src: wmImg
  }
];

export interface HomeProps {
  fetchProductsByType: typeof fetchProductsByType;
}

export const _Home: React.FC<HomeProps> = ({ fetchProductsByType }) => {
  return (
    <Wrapper>
      {categories.map(({ title, src }) => (
        <NavLink
          onClick={() => fetchProductsByType(title.toLowerCase() as Type)}
          to="/shop"
          key={title}
        >
          <NavTitle>{title}</NavTitle>
          <ImageWrapper>
            <NavImage src={src} alt="" />
          </ImageWrapper>
        </NavLink>
      ))}
    </Wrapper>
  );
};

export const Home = connect(
  null,
  { fetchProductsByType }
)(_Home);
