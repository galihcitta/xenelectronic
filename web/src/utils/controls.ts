import { Controls } from '../pages/Auth/types';

export const signInControls: Controls = {
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signin-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signin-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};

export const signUpControls: Controls = {
  username: {
    label: 'Enter username',
    placeholder: '',
    type: 'text',
    id: 'signup-username',
    name: 'username',
    value: '',
    validationRules: {
      required: true,
      minLength: 4,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signup-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signup-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true,
      isSecure: true,
      minLength: 8,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};
