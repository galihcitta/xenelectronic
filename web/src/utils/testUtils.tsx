import React from 'react';
import { BrowserRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../modules/store';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

export const withRouter = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <BrowserRouter>
    <Component {...props} />
  </BrowserRouter>
);

export const withHistory = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Router history={history}>
    <Component {...props} />
  </Router>
);

export const withProvider = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Provider store={store}>
    <BrowserRouter>
      <Component {...props} />
    </BrowserRouter>
  </Provider>
);
