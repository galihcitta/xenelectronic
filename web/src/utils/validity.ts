import { ValidationRules } from '../pages/Auth/types';

export const validate = (
  value: string,
  controlName: string,
  rules: ValidationRules
) => {
  let isValid = true;
  let errorMsg = '';

  if (rules.required) {
    isValid = value.trim() !== '' && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} field is required`;
  }

  if (rules.minLength) {
    isValid = value.length >= rules.minLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be at least ${rules.minLength} characters`;
  }

  if (rules.maxLength) {
    isValid = value.length <= rules.maxLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be no longer than ${rules.maxLength} characters`;
  }

  if (rules.isEmail) {
    const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `email is invalid`;
  }

  if (rules.isSecure) {
    // Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character
    const pattern = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `password too weak`;
  }

  return {
    isValid,
    errorMsg: errorMsg.charAt(0).toUpperCase() + errorMsg.slice(1)
  };
};
