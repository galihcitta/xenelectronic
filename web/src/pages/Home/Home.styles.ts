import styled from 'styled-components';
import { Link } from 'react-router-dom';

import { device, fontWeight, color } from '../../utils/styles';

export const Wrapper = styled.section`
  width: 100%;
  display: flex;
  flex-direction: column;
  max-width: 500px;
  margin: 0 auto;

  @media ${device.tablet} {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-column-gap: 2em;
    max-width: 100%;
  }

  @media ${device.laptop} {
    padding-top: 1.2em;
  }
`;

export const NavTitle = styled.h2`
  position: absolute;
  z-index: 2;
  top: 70%;
  left: 0;
  color: ${color.black};
  font-weight: ${fontWeight.bold};
  font-size: 1.15rem;
  background: ${color.white};
  padding: 0.3em 2em 0.3em 0.5em;

  @media ${device.mobileL} {
    font-size: 1.25rem;
  }
`;

export const ImageWrapper = styled.div`
  width: 100%;
  max-height: 260px;
  overflow: hidden;
  position: relative;
  transition: opacity 0.3s ease-in-out;

  @media ${device.mobileL} {
    max-height: 360px;
  }

  @media ${device.tablet} {
    height: calc(20px + 29vw);
  }

  @media ${device.laptop} {
    max-height: 350px;
  }

  &:after {
    content: '';
    transition: all 1.4s ease-in-out;
    position: absolute;
    width: 100%;
    height: 100%;
    top: 0;
    right: 0;
    background: ${color.white};
    animation: slideIn 1.4s forwards;
  }

  @keyframes slideIn {
    0% {
      width: 100%;
    }
    100% {
      width: 0%;
    }
  }
`;

export const NavLink = styled(Link)`
  position: relative;
  display: block;
  margin-bottom: 1.4em;

  &:hover ${ImageWrapper} {
    opacity: 0.45;
  }
`;

export const NavImage = styled.img`
  width: 100%;
  transform: scale(1.3);
  transition: transform 1.4s ease-in-out;
  animation: zoomOut 1.4s forwards;

  @keyframes zoomOut {
    0% {
      transform: scale(1.3);
    }
    100% {
      transform: scale(1);
    }
  }
`;
