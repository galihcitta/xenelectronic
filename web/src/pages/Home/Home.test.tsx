import { render, fireEvent } from '@testing-library/react';
import { HomeProps, _Home } from './Home.component';
import { withRouter } from '../../utils/testUtils';

const HomeWithProvider = withRouter(_Home);

describe('<Home />', () => {
  const mockFetchProductsByType = jest.fn();
  const props: HomeProps = {
    fetchProductsByType: mockFetchProductsByType
  };

  const setup = (props: HomeProps) => {
    const utils = render(HomeWithProvider(props));
    return { ...utils };
  };

  it('renders correctly', () => {
    const { asFragment } = setup(props);
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls `fetchProductsByType` action', () => {
    const { getByText } = setup(props);
    fireEvent.click(getByText('Clocks'));

    expect(mockFetchProductsByType).toBeCalledTimes(1);
    expect(mockFetchProductsByType).toBeCalledWith('clocks');
  });
});
