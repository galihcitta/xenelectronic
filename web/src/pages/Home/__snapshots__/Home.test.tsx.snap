// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders correctly 1`] = `
<DocumentFragment>
  <section
    class="sc-bdVaJa heQTnx"
  >
    <a
      class="sc-bxivhb bMgZix"
      href="/shop"
    >
      <h2
        class="sc-bwzfXH cVaPpq"
      >
        Lamps
      </h2>
      <div
        class="sc-htpNat fCRfDq"
      >
        <img
          alt=""
          class="sc-ifAKCX VXSCg"
          src="lamp.png"
        />
      </div>
    </a>
    <a
      class="sc-bxivhb bMgZix"
      href="/shop"
    >
      <h2
        class="sc-bwzfXH cVaPpq"
      >
        Vases
      </h2>
      <div
        class="sc-htpNat fCRfDq"
      >
        <img
          alt=""
          class="sc-ifAKCX VXSCg"
          src="vase.png"
        />
      </div>
    </a>
    <a
      class="sc-bxivhb bMgZix"
      href="/shop"
    >
      <h2
        class="sc-bwzfXH cVaPpq"
      >
        Chairs
      </h2>
      <div
        class="sc-htpNat fCRfDq"
      >
        <img
          alt=""
          class="sc-ifAKCX VXSCg"
          src="chair.png"
        />
      </div>
    </a>
    <a
      class="sc-bxivhb bMgZix"
      href="/shop"
    >
      <h2
        class="sc-bwzfXH cVaPpq"
      >
        Clocks
      </h2>
      <div
        class="sc-htpNat fCRfDq"
      >
        <img
          alt=""
          class="sc-ifAKCX VXSCg"
          src="clock.png"
        />
      </div>
    </a>
    <a
      class="sc-bxivhb bMgZix"
      href="/shop"
    >
      <h2
        class="sc-bwzfXH cVaPpq"
      >
        Benches
      </h2>
      <div
        class="sc-htpNat fCRfDq"
      >
        <img
          alt=""
          class="sc-ifAKCX VXSCg"
          src="bench.png"
        />
      </div>
    </a>
    <a
      class="sc-bxivhb bMgZix"
      href="/shop"
    >
      <h2
        class="sc-bwzfXH cVaPpq"
      >
        Flowerpots
      </h2>
      <div
        class="sc-htpNat fCRfDq"
      >
        <img
          alt=""
          class="sc-ifAKCX VXSCg"
          src="flowerpot.png"
        />
      </div>
    </a>
    <a
      class="sc-bxivhb bMgZix"
      href="/shop"
    >
      <h2
        class="sc-bwzfXH cVaPpq"
      >
        Plants
      </h2>
      <div
        class="sc-htpNat fCRfDq"
      >
        <img
          alt=""
          class="sc-ifAKCX VXSCg"
          src="plant.png"
        />
      </div>
    </a>
  </section>
</DocumentFragment>
`;
